mission_control/src/mission_control.cpp

```diff
@@ -34,165 +34,95 @@
-  pub_report_heartbeat =
-      node_handle.advertise<mission_control::ReportHeartbeat>("/mngr/report_heartbeat", 100);
+  pub_report_mission_load_state =
+      pnh.advertise<mission_control::ReportLoadMissionState>("report_mission_load_state", 100);
+  pub_report_mission_execute_state = pnh.advertise<mission_control::ReportExecuteMissionState>(
+      "report_mission_execute_state", 100);
+  pub_report_missions = pnh.advertise<mission_control::ReportMissions>("report_missions", 100);
+  pub_report_heartbeat = pnh.advertise<mission_control::ReportHeartbeat>("report_heartbeat", 100);
 
   reportExecuteMissionStateTimer =
-      node_handle.createTimer(ros::Duration(reportExecuteMissionStateRate),
-                              &MissionControlNode::reportExecuteMissionState, this);
-  reportHeartbeatTimer = node_handle.createTimer(ros::Duration(reportHeartbeatRate),
-                                                 &MissionControlNode::reportHeartbeat, this);
+      pnh.createTimer(ros::Duration(reportExecuteMissionStateRate),
+                      &MissionControlNode::reportExecuteMissionState, this);
+  reportHeartbeatTimer = pnh.createTimer(ros::Duration(reportHeartbeatRate),
+                                         &MissionControlNode::reportHeartbeat, this);
+
+  executeMissionTimer =
+      pnh.createTimer(ros::Duration(0.1), &MissionControlNode::executeMissionT, this);
+  executeMissionTimer.stop();
 }
 
-MissionControlNode::~MissionControlNode()
-{
-  stop();
-
-  std::map<int, Mission*>::iterator it;
-
-  for (it = m_mission_map.begin(); it != m_mission_map.end(); it++)
-  {
-    if (it->second != NULL)
-    {
-      delete it->second;
-    }
-  }
-
-  m_mission_map.clear();
-}
+MissionControlNode::~MissionControlNode() { m_mission_map.clear(); }
 
 int MissionControlNode::loadMissionFile(std::string mission_full_path)
 {
-  MissionParser parser(node_handle);
-
-  Mission* newMission = new Mission();
-  // Parse the specified mission
-  if (parser.parseMissionFile(*newMission, mission_full_path) == false)
+  std::unique_ptr<Mission> newMission = std::make_unique<Mission>();
+  if (newMission->loadMission(mission_full_path) == true)
   {
-    delete newMission;
-    return -1;
+    m_mission_id_counter++;
+    m_mission_map[m_mission_id_counter] = std::move(newMission);
+    return 0;
   }
-
-  m_mission_id_counter++;
-
-  m_mission_map[m_mission_id_counter] = newMission;
-
-  return 0;
-}
-
-int MissionControlNode::executeMission(int missionId)
-{
-  if ((missionId > 0) && (m_mission_map.size() > 0) && (m_mission_map.count(missionId) > 0))
+  else
   {
-    last_state = Mission::MissionState::READY;
-    last_id = -1;
-    m_current_mission_id = missionId;
-    m_mission_map[m_current_mission_id]->ExecuteMission(node_handle);
+    return -1;
   }
 }
 
 int MissionControlNode::abortMission(int missionId)
 {
-  if ((missionId > 0) && (m_mission_map.size() > 0) && (m_mission_map.count(missionId) > 0))
+  executeMissionTimer.stop();
```

Comment:
@patapufate should the timer be stopped if and only if `m_current_mission_id == missionId`?

Reply:
yes. Fixed